domtrip normalizes CRLF line endings to LF when writing XML

## Description

When domtrip modifies an XML file that uses CRLF (`\r\n`) line endings, the written output converts the modified sections to LF (`\n`) while the rest of the file retains CRLF. This creates a mixed-endings file that fails tools like Spotless that enforce consistent line endings.

## Steps to Reproduce

1. Have a POM file with CRLF line endings (common on Windows or in projects using CRLF by convention)
2. Use domtrip (via `mvnup apply`) to modify the POM (e.g., apply inference optimizations, upgrade plugins)
3. The modified sections of the POM now have LF line endings while untouched sections keep CRLF

## Observed Behavior

The Spotless diff from `sling-org-apache-sling-testing-hamcrest` shows:

```
[ERROR]     pom.xml
[ERROR]         @@ -17,67 +17,67 @@
[ERROR]            specific language governing permissions and limitations\r\n
[ERROR]            under the License.\r\n
[ERROR]          -->\r\n
[ERROR]         -<project xmlns="http://maven.apache.org/POM/4.0.0" ...>\n
[ERROR]         -\n
[ERROR]         -    <modelVersion>4.0.0</modelVersion>\n
```

The original file has `\r\n` throughout, but after domtrip modifies it, the `<project>` element and its contents use `\n`.

## Expected Behavior

domtrip should detect and preserve the original line ending style of the file when writing output.

## Impact

This breaks builds for projects that use Spotless, Checkstyle, or other formatting tools that enforce consistent line endings. Found during Maven 4 compatibility testing of ~960 Apache projects via `mvnup apply`.

_Claude Code on behalf of Guillaume Nodet_

--- a/core/src/main/java/eu/maveniverse/domtrip/Editor.java
+++ b/core/src/main/java/eu/maveniverse/domtrip/Editor.java
@@ -363,7 +363,7 @@ public class Editor {
 
     private void handleOnlyElementRemoval(ContainerNode container) {
         if (container instanceof Element) {
-            ((Element) container).innerPrecedingWhitespace("\n");
+            ((Element) container).innerPrecedingWhitespace(lineEnding);
         }
     }
 
@@ -404,7 +404,7 @@ public class Editor {
         String[] lines = whitespace.split("\\r?\\n", -1);
         if (lines.length > 1) {
             // Return newline + last line (which should be the indentation)
-            return "\n" + lines[lines.length - 1];
+            return lineEnding + lines[lines.length - 1];
         } else {
             // No newlines, return as-is
             return whitespace;
@@ -431,7 +431,7 @@ public class Editor {
         String indentation = lines[lines.length - 1];
 
         // Return just newline + indentation, removing any extra blank lines
-        return "\n" + indentation;
+        return lineEnding + indentation;
     }
 
     /**
@@ -1681,7 +1681,7 @@ public class Editor {
         }
 
         // Extract the pattern after the last newline (including the newline)
-        int lastNewline = existingWhitespace.lastIndexOf(lineEnding.charAt(lineEnding.length() - 1));
+        int lastNewline = existingWhitespace.lastIndexOf(lineEnding);
         if (lastNewline >= 0) {
             // Return from the newline onwards (including the newline)
             return existingWhitespace.substring(lastNewline);

--- a/core/src/test/java/eu/maveniverse/domtrip/CrlfPreservationTest.java
+++ b/core/src/test/java/eu/maveniverse/domtrip/CrlfPreservationTest.java
@@ -1,0 +1,182 @@
+package eu.maveniverse.domtrip;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests that CRLF line endings are preserved during XML modifications.
+ *
+ * @see <a href="https://github.com/maveniverse/domtrip/issues/230">Issue #230</a>
+ */
+class CrlfPreservationTest {
+
+    private static String crlfXml(String... lines) {
+        return String.join("\r\n", lines);
+    }
+
+    @Test
+    void testRoundTripPreservesCrlf() throws Exception {
+        String xml = crlfXml(
+                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>",
+                "<project>",
+                "    <groupId>com.example</groupId>",
+                "    <artifactId>test</artifactId>",
+                "    <version>1.0</version>",
+                "</project>",
+                "");
+
+        Document doc = Document.of(xml);
+        Editor editor = new Editor(doc);
+        String result = editor.toXml();
+
+        assertEquals(xml, result);
+    }
+
+    @Test
+    void testModifyTextPreservesCrlf() throws Exception {
+        String xml = crlfXml(
+                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>",
+                "<project>",
+                "    <groupId>com.example</groupId>",
+                "    <artifactId>test</artifactId>",
+                "    <version>1.0</version>",
+                "</project>",
+                "");
+
+        Document doc = Document.of(xml);
+        Editor editor = new Editor(doc);
+        Element version = doc.root().childElement("version").orElseThrow();
+        version.textContent("2.0");
+
+        String result = editor.toXml();
+
+        assertFalse(result.contains("\r\n2.0"), "Modified text should not introduce bare LF before value");
+        assertTrue(result.contains("2.0"), "Modified text should be present");
+        assertNoCrlfCorruption(result);
+    }
+
+    @Test
+    void testAddElementPreservesCrlf() throws Exception {
+        String xml = crlfXml(
+                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>",
+                "<project>",
+                "    <groupId>com.example</groupId>",
+                "    <artifactId>test</artifactId>",
+                "</project>",
+                "");
+
+        Document doc = Document.of(xml);
+        Editor editor = new Editor(doc);
+        editor.addElement(doc.root(), "version", "1.0");
+
+        String result = editor.toXml();
+
+        assertTrue(result.contains("<version>1.0</version>"));
+        assertNoCrlfCorruption(result);
+    }
+
+    @Test
+    void testRemoveElementPreservesCrlf() throws Exception {
+        String xml = crlfXml(
+                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>",
+                "<project>",
+                "    <groupId>com.example</groupId>",
+                "    <artifactId>test</artifactId>",
+                "    <version>1.0</version>",
+                "</project>",
+                "");
+
+        Document doc = Document.of(xml);
+        Editor editor = new Editor(doc);
+        Element version = doc.root().childElement("version").orElseThrow();
+        editor.removeElement(version);
+
+        String result = editor.toXml();
+
+        assertFalse(result.contains("<version>"));
+        assertNoCrlfCorruption(result);
+    }
+
+    @Test
+    void testRemoveOnlyChildPreservesCrlf() throws Exception {
+        String xml = crlfXml(
+                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>",
+                "<project>",
+                "    <dependencies>",
+                "        <dependency>",
+                "            <groupId>junit</groupId>",
+                "        </dependency>",
+                "    </dependencies>",
+                "</project>",
+                "");
+
+        Document doc = Document.of(xml);
+        Editor editor = new Editor(doc);
+        Element deps = doc.root().childElement("dependencies").orElseThrow();
+        Element dep = deps.childElement("dependency").orElseThrow();
+        editor.removeElement(dep);
+
+        String result = editor.toXml();
+
+        assertFalse(result.contains("<dependency>"));
+        assertNoCrlfCorruption(result);
+    }
+
+    @Test
+    void testAddAttributePreservesCrlf() throws Exception {
+        String xml = crlfXml(
+                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>",
+                "<project>",
+                "    <groupId>com.example</groupId>",
+                "</project>",
+                "");
+
+        Document doc = Document.of(xml);
+        Editor editor = new Editor(doc);
+        editor.setAttribute(doc.root().childElement("groupId").orElseThrow(), "scope", "test");
+
+        String result = editor.toXml();
+
+        assertTrue(result.contains("scope"));
+        assertNoCrlfCorruption(result);
+    }
+
+    @Test
+    void testMultipleModificationsPreserveCrlf() throws Exception {
+        String xml = crlfXml(
+                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>",
+                "<project>",
+                "    <groupId>com.example</groupId>",
+                "    <artifactId>test</artifactId>",
+                "    <version>1.0</version>",
+                "    <packaging>jar</packaging>",
+                "</project>",
+                "");
+
+        Document doc = Document.of(xml);
+        Editor editor = new Editor(doc);
+
+        doc.root().childElement("version").orElseThrow().textContent("2.0");
+        editor.removeElement(doc.root().childElement("packaging").orElseThrow());
+        editor.addElement(doc.root(), "name", "My Project");
+
+        String result = editor.toXml();
+
+        assertTrue(result.contains("2.0"));
+        assertFalse(result.contains("<packaging>"));
+        assertTrue(result.contains("<name>My Project</name>"));
+        assertNoCrlfCorruption(result);
+    }
+
+    private void assertNoCrlfCorruption(String xml) {
+        for (int i = 0; i < xml.length(); i++) {
+            if (xml.charAt(i) == '\n' && (i == 0 || xml.charAt(i - 1) != '\r')) {
+                int start = Math.max(0, i - 30);
+                int end = Math.min(xml.length(), i + 30);
+                String context = xml.substring(start, end).replace("\r", "\\r").replace("\n", "\\n");
+                fail("Found bare LF (without CR) at position " + i + ", context: ..." + context + "...");
+            }
+        }
+    }
+}

--- a/core/src/test/java/eu/maveniverse/domtrip/CrlfPreservationTest.java
+++ b/core/src/test/java/eu/maveniverse/domtrip/CrlfPreservationTest.java
@@ -51,7 +51,7 @@ class CrlfPreservationTest {
 
         String result = editor.toXml();
 
-        assertFalse(result.contains("\r\n2.0"), "Modified text should not introduce bare LF before value");
+        assertFalse(result.contains("\n2.0"), "Modified text should not introduce bare LF before value");
         assertTrue(result.contains("2.0"), "Modified text should be present");
         assertNoCrlfCorruption(result);
     }

--- a/core/src/test/java/eu/maveniverse/domtrip/CrlfPreservationTest.java
+++ b/core/src/test/java/eu/maveniverse/domtrip/CrlfPreservationTest.java
@@ -143,6 +143,23 @@ class CrlfPreservationTest {
     }
 
     @Test
+    void testRemoveMiddleElementWithInlineNextPreservesCrlf() throws Exception {
+        String xml = "<root>\r\n  <a/>\r\n  <b/> <c/>\r\n</root>";
+
+        Document doc = Document.of(xml);
+        Editor editor = new Editor(doc);
+        Element b = doc.root().childElement("b").orElseThrow();
+        editor.removeElement(b);
+
+        String result = editor.toXml();
+
+        assertTrue(result.contains("<a/>"));
+        assertTrue(result.contains("<c/>"));
+        assertFalse(result.contains("<b/>"));
+        assertNoCrlfCorruption(result);
+    }
+
+    @Test
     void testMultipleModificationsPreserveCrlf() throws Exception {
         String xml = crlfXml(
                 "<?xml version=\"1.0\" encoding=\"UTF-8\"?>",

